### PR TITLE
邮件和短域名

### PR DESCRIPTION
**Pull request by nothatDinger**
邮件加了抄送和密抄，接口有相应的增加改变

加了个短域名服务，
生成短链的接口为/genurl
需要传入带urllist字段，值为url数组的json

解析短链的接口为/v/:shorturl

### CODE REVIEW

**Comment by wangnengjie** [comment]
LGTM. I give you the write permission. You can push to this repo if you have any update commits laster. You can use `sync.waitGroup` and channel with cache to make your code better. 
And I wonder what's the use for `ShortUrl` and `HashCode` fields in struct `shorter`?
https://github.com/nothatDinger/open-platform/blob/a7e417719f160527fc53aa029bef753705111e6d/handler/genshorturl.go#L26

**Comment by nothatDinger** [comment]
Actually they are useless ,I'll delete them in update commit soon



---原始邮件---
发件人: "Panda"<notifications@github.com&gt;
发送时间: 2021年1月21日(周四) 晚上9:27
收件人: "UniqueStudio/open-platform"<open-platform@noreply.github.com&gt;;
抄送: "nothatDinger"<578279246@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题: Re: [UniqueStudio/open-platform] 邮件和短域名 (#7)




 
LGTM. I give you the write permission. You can push to this repo if you have any update commits laster. You can use sync.waitGroup and channel with cache to make your code better.
 And I wonder what's the use for ShortUrl and HashCode fields in struct shorter?
 https://github.com/nothatDinger/open-platform/blob/a7e417719f160527fc53aa029bef753705111e6d/handler/genshorturl.go#L26

File: db/mysql.go [hunk 1]
```diff
@@ -20,7 +21,7 @@ func init() {
 		log.Println(err)
 		panic(err)
 	}
-	err = ORM.Sync(new(Reply), new(Status))
+	err = ORM.Sync(new(Reply), new(Status),new(Short_Url))
```
Comment (ttzztztz) [inline db/mysql.go:24]: one more `backspace`, pay attention to the code-style

File: handler/genshorturl.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+	"open-platform/db"
+	"open-platform/utils"
+
+	"github.com/gin-gonic/gin"
+)
+
+type resultForCreation struct {
+	url  string
+	code string
+	err error
```
Comment (ttzztztz) [inline handler/genshorturl.go:18]: one more `backspace`, pay attention to the code-style.

You can use the code format tool in the IDE

File: handler/genshorturl.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+
+type shorter struct {
+
```
Comment (ttzztztz) [inline handler/genshorturl.go:23]: remove this line

File: handler/genshorturl.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	}
+
+	var results = make(map[string]interface{})
+	var count = 1
+	for {
+		res := <-resultCode
+
+		if res.err !=nil{
+			results[res.url] = res.err.Error()
+		}else{
```
Comment (ttzztztz) [inline handler/genshorturl.go:65]: code-style

File: handler/genshorturl.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+			results[res.url] = shortUrlPrefix + res.code
+		}
+
+		if count == len(urlList) {
```
Comment (ttzztztz) [inline handler/genshorturl.go:69]: use `sync.WaitGroup` in the golang standard library instead

File: handler/mail.go [hunk 6]
```diff
@@ -26,6 +28,8 @@ func SendMailHandler(c *gin.Context) {
 	var data sender
 	c.BindJSON(&data)
 
+	bccto :=data.BccTo
```
Comment (ttzztztz) [inline handler/mail.go:31]: pay attention to the code style. One more `backspace`

File: handler/mail.go [hunk 7]
```diff
@@ -26,6 +28,8 @@ func SendMailHandler(c *gin.Context) {
...
+	ccto :=data.CcTo
```
Comment (ttzztztz) [inline handler/mail.go:32]: pay attention to the code style. One more `backspace`

File: handler/mail.go [hunk 8]
```diff
@@ -49,10 +53,10 @@ func SendMailHandler(c *gin.Context) {
 		utils.AppConfig.SMTP.Sender,
 		utils.AppConfig.SMTP.Password,
 		utils.AppConfig.SMTP.Host,
-		subject, renderContent, "html", utils.RemoveDuplicate(append(toList, to)))
+		subject, renderContent, "html", utils.RemoveDuplicate(append(toList, to)),ccto,bccto)
```
Comment (ttzztztz) [inline handler/mail.go:56]: pay attention to the code style. Two more `backspace`

File: handler/mail.go [hunk 9]
```diff
@@ -62,7 +66,7 @@ func SendMailHandler(c *gin.Context) {
 }
 
 // SendToMail is a function to handle send email smtp requests
-func SendToMail(user, password, host, subject, body, mailtype string, to []string) error {
+func SendToMail(user, password, host, subject, body, mailtype string, to,ccto,bccto []string) error {
```
Comment (ttzztztz) [inline handler/mail.go:69]: pay attention to the code style. Two more `backspace`

File: handler/mail.go [hunk 10]
```diff
@@ -82,3 +92,10 @@ func encodeRFC2047(String string) string {
 	addr := mail.Address{Name: String, Address: ""}
 	return strings.Trim(addr.String(), "<@>")
 }
+
+func MergeSlice(s1 []string, s2 []string) []string {
```
Comment (ttzztztz) [inline handler/mail.go:96]: Use `append` method in the standard library and `...` operator instead.
You can find references at https://www.geeksforgeeks.org/how-to-append-a-slice-in-golang/

Don't write a method that already exists in the standard library

File: handler/shorturl.go [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+import (
+	"github.com/gin-gonic/gin"
+	"net/http"
+	"open-platform/db"
+	"strings"
+)
+
+func MapShorturlHandler(c *gin.Context)  {
+	var result db.Short_Url
+	shorturl :=  c.Param("Shorturl")
```
Comment (ttzztztz) [inline handler/shorturl.go:12]: code style

File: handler/shorturl.go [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+	has,err:=db.ORM.Where("Shorturl=?",shorturl).Get(&result)
```
Comment (ttzztztz) [inline handler/shorturl.go:13]: code style

File: handler/shorturl.go [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+	if err !=nil {
```
Comment (ttzztztz) [inline handler/shorturl.go:15]: code style

File: main.go [hunk 14]
```diff
@@ -95,6 +95,9 @@ func main() {
 		open.GET("/permission", handler.GetPermissionHandler)
 	}
 
+	r.GET("/v/:Shorturl",handler.MapShorturlHandler)
+	r.GET("/genUrl",middleware.Auth(),handler.CreateShortUrlHandler)
```
Comment (ttzztztz) [inline main.go:99]: code style

File: utils/config.go [hunk 15]
```diff
@@ -5,54 +5,53 @@ import (
...
 type Config struct {
 	APPName string `default:"Gin App"`
-
+	ShortUrl string `default:"uniqs.cc"`
 	Server struct {
-		Host      string `default:"127.0.0.1"`
-		Hostname  string `default:"open.hustunique.com"`
-		Port      string `default:"9012"`
-		SecretKey string `default:"SecretKey"`
+		Host      string `default:"127.0.0.1";yaml:"host"`
```
Comment (ttzztztz) [inline utils/config.go:18]: The spliter of golang struct tag is ` ` (backspace), not `;`

Reference
https://medium.com/golangspec/tags-in-golang-3e5db0b8ef3e
https://stackoverflow.com/questions/18635671/how-to-define-multiple-name-tags-in-a-struct

**Review by ttzztztz — CHANGES_REQUESTED** [review]

**Comment by wangnengjie** [comment]
Why a binary file "open-platform" here?
https://github.com/UniqueStudio/open-platform/pull/7/files#diff-ef0c3328876d5e5355d13903a4e2a2f618c85da9f4e3cd652cbfebf3a9db57ef

File: utils/config.go [hunk 16]
```diff
@@ -5,54 +5,53 @@ import (
...
 type Config struct {
 	APPName string `default:"Gin App"`
-
+	ShortUrl string `default:"uniqs.cc"`
 	Server struct {
-		Host      string `default:"127.0.0.1"`
-		Hostname  string `default:"open.hustunique.com"`
-		Port      string `default:"9012"`
-		SecretKey string `default:"SecretKey"`
+		Host      string `default:"127.0.0.1",yaml:"host"`
```
Comment (ttzztztz) [inline utils/config.go:18]: `default:"127.0.0.1" yaml:"host"`

**Review by ttzztztz — CHANGES_REQUESTED** [review]

File: utils/config.go (hunk 16 above)
Comment (ttzztztz) [reply utils/config.go:18]: Other similarities

Comment (nothatDinger) [reply utils/config.go:18]: ok

**Pull request merged**